{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions and Imports for Ovarian Cancer Ratio Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to cptac, a python package for disseminating cancer\n",
      "proteogenomics data. To view available datasets, enter\n",
      "'cptac.list_data()'. Extensive tutorials are available at\n",
      "https://github.com/PayneLab/cptac/tree/master/doc\n",
      "\n",
      "******\n",
      "Version: 0.4.5\n",
      "******\n",
      "Checking that data files are up-to-date...\n",
      "\n",
      "Data check complete.\n",
      "ovarian data version: Most recent release\n",
      "\n",
      "Loading clinical data...\n",
      "Loading cnv data...\n",
      "Loading definitions data...\n",
      "Loading phosphoproteomics data...\n",
      "Loading proteomics data...\n",
      "Loading somatic_38 data...\n",
      "Loading transcriptomics data...\n",
      "Loading treatment data...\n",
      "Checking that data files are up-to-date...\n",
      "\n",
      "Data check complete.\n",
      "colon data version: Most recent release\n",
      "\n",
      "Loading clinical data...\n",
      "Loading miRNA data...\n",
      "Loading mutation data...\n",
      "Loading mutation_binary data...\n",
      "Loading phosphoproteomics_normal data...\n",
      "Loading phosphoproteomics_tumor data...\n",
      "Loading proteomics_normal data...\n",
      "Loading proteomics_tumor data...\n",
      "Loading transcriptomics data...\n",
      "Checking that data files are up-to-date...\n",
      "\n",
      "Data check complete.\n",
      "endometrial data version: 2.1\n",
      "\n",
      "Loading acetylproteomics data...\n",
      "Loading clinical data...\n",
      "Loading CNA data...\n",
      "Loading definitions data...\n",
      "Loading miRNA data...\n",
      "Loading phosphoproteomics_gene data...\n",
      "Loading phosphoproteomics_site data...\n",
      "Loading proteomics data...\n",
      "Loading somatic data...\n",
      "Loading somatic_binary data...\n",
      "Loading transcriptomics_circular data...\n",
      "Loading transcriptomics_linear data...\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import cptac\n",
    "ov = cptac.Ovarian()\n",
    "colon = cptac.Colon()\n",
    "en = cptac.Endometrial()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import patient data as well as a list of protein complexes from CORUM (http://mips.helmholtz-muenchen.de/corum/#download 'Complete Complexes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv('/Users/amandaoliphant/Documents/GitHub/OvarianProspective/Protein_Complexes/Data_Files/proteinGroups_cleaned.txt', sep='\\t')\n",
    "ov_clinical = ov.get_clinical()\n",
    "ov_mutations = ov.get_mutations()\n",
    "colon_clinical = colon.get_clinical()\n",
    "colon_mutations = colon.get_mutations()\n",
    "en_clinical = en.get_clinical()\n",
    "en_mutations = en.get_mutations()\n",
    "\n",
    "# Remove duplicate rows from the dataset\n",
    "data = data.set_index('Gene_Name')\n",
    "data = data.loc[~data.index.duplicated()]\n",
    "complexData = pd.read_csv('/Users/amandaoliphant/Documents/GitHub/OvarianProspective/Protein_Complexes/Data_Files/allComplexes.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import transcriptional regulatory interactions from TRRUST v2 (https://www.grnpedia.org/trrust/downloadnetwork.php) and mutation data for tumor samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "ov_mutation_data = ov_clinical[['Patient_ID', 'Sample_Tumor_Normal']].join(ov_mutations).dropna(axis = 0)\n",
    "ov_mutation_data = ov_mutation_data.loc[ov_mutation_data['Sample_Tumor_Normal'] == 'Tumor']\n",
    "\n",
    "colon_mutation_data = colon_clinical[['Patient_ID', 'Sample_Tumor_Normal']].join(colon_mutations).dropna(axis = 0)\n",
    "colon_mutation_data = colon_mutation_data.loc[colon_mutation_data['Sample_Tumor_Normal'] == 'Tumor']\n",
    "\n",
    "en_mutation_data = en_clinical[['Patient_ID', 'Proteomics_Tumor_Normal']].join(en_mutations).dropna(axis = 0)\n",
    "en_mutation_data = en_mutation_data.loc[en_mutation_data['Proteomics_Tumor_Normal'] == 'Tumor']\n",
    "\n",
    "tf_data = pd.read_csv('/Users/amandaoliphant/Documents/GitHub/OvarianProspective/Protein_Complexes/Data_Files/trrust_rawdata.human.tsv', sep='\\t', header=None)\n",
    "tf_data.columns = ['TF', 'Regulated_Gene', 'Regulation_Type', 'ID']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create a workable dictionary of protein complex information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select for human complexes\n",
    "complexData = complexData.loc[complexData['Organism'] == 'Human']\n",
    "\n",
    "# Split the proteins in each complex into invdividual columns of a new df\n",
    "subunitNames = complexData['subunits(Gene name)'].str.split(';', expand = True)\n",
    "subunitNames.index = complexData.ComplexName\n",
    "\n",
    "# Create a dictionary (key = complex name, value = list of proteins in complex) and remove None values\n",
    "subunitNames = subunitNames.loc[~subunitNames.index.duplicated()]\n",
    "subunitNames = subunitNames.transpose().to_dict('list')\n",
    "for key, val in subunitNames.items():\n",
    "    subunitNames[key] = [value for value in val if value != None]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define basic analysis and plotting functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get_ratio_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "get_ratio_df\n",
    "-----------\n",
    "All patients, single protein, tumor vs normal\n",
    "Returns two dataframes: tumor and normal containing ratios for the proteins\n",
    "\n",
    "Parameters:\n",
    "    prot1, prot2 = (gene) names of the two proteins to be compared (ratio of prot1 / prot2)\n",
    "    \n",
    "Returns:\n",
    "    Two dataframes: tumor and normal containing ratios for the proteins\n",
    "  \n",
    "\"\"\"\n",
    "def get_ratio_df(prot1, prot2):\n",
    "    # Make sure that both proteins are in our dataset\n",
    "    if not data.index.contains(prot1) or not data.index.contains(prot2): return None, None\n",
    "    \n",
    "    # Create a dataframe of ratios for easy testing\n",
    "    tumor_rows_list = []\n",
    "    normal_rows_list = []\n",
    "    for patient in data.columns:\n",
    "        \n",
    "        # Determine the sample type\n",
    "        sample_type = 'Tumor'\n",
    "        if '_NM' in patient: sample_type = 'Normal'\n",
    "            \n",
    "        # Find the ratio prot1 / prot2 and create a new row for the dataframe\n",
    "        if data.at[prot2, patient] == 0: \n",
    "            ratio = np.nan\n",
    "        else: \n",
    "            ratio = data.at[prot1, patient] / data.at[prot2, patient]\n",
    "        # Create a row entry for the dictionary\n",
    "        row_dict = {'Ratio': ratio, 'Sample_Type': sample_type, 'Patient_ID': patient}\n",
    "        \n",
    "        # Add the new row to the tumor or normal list, depending on the sample type\n",
    "        if sample_type == 'Tumor':\n",
    "            tumor_rows_list.append(row_dict)\n",
    "        else:\n",
    "            normal_rows_list.append(row_dict)\n",
    "    \n",
    "    # Convert the row lists into dataframes\n",
    "    tumor_ratio_df = pd.DataFrame(tumor_rows_list)\n",
    "    normal_ratio_df = pd.DataFrame(normal_rows_list)\n",
    "    \n",
    "    # If there were no valid ratios for either the tumor or normal samples, return None\n",
    "    if (not np.isfinite(tumor_ratio_df['Ratio']).any()) or (not np.isfinite(normal_ratio_df['Ratio']).any()): \n",
    "        return None, None\n",
    "    \n",
    "    # Drop na values\n",
    "    tumor_ratio_df = tumor_ratio_df.dropna(axis = 0)\n",
    "    normal_ratio_df = normal_ratio_df.dropna(axis = 0)\n",
    "    \n",
    "    return tumor_ratio_df, normal_ratio_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### test_complex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "test_complex\n",
    "------------\n",
    "Perform statistical tests on every combination of proteins in a given complex, printing those with significant p-values\n",
    "\n",
    "Parameters:\n",
    "    complex_name = name of the complex\n",
    "    test_type = type of test to perform, levene or ttest\n",
    "    \n",
    "Returns:\n",
    "    Classification of the complex: whether it has proteins missing in the normal, missing proteins in cancer,\n",
    "    or other (proteins present in both)\n",
    "    Prints tests with statistically significant values\n",
    "\n",
    "\"\"\"\n",
    "def test_complex(complex_name, test_type = 'ttest'):\n",
    "    \n",
    "    prot_list = subunitNames[complex_name]\n",
    "    sig_result_found = False\n",
    "    \n",
    "    # Keep track of how many significant ratios in this complex include only 0 values for tumor/normal \n",
    "    num_missing_tumor = 0\n",
    "    num_missing_normal = 0\n",
    "    \n",
    "    # Calculate the cutoff for significance (Bonferroni correction)\n",
    "    alpha = 0.05 / (len(subunitNames)*(len(subunitNames) - 1))\n",
    "    \n",
    "    for i in range(0, len(prot_list)):\n",
    "        for j in range(0, len(prot_list)):\n",
    "            if j == i: continue\n",
    "                \n",
    "            # Test the two proteins (ratio of prot1 / prot2)\n",
    "            tumor_ratio_df, normal_ratio_df = get_ratio_df(prot_list[i], prot_list[j])\n",
    "            if tumor_ratio_df is None or normal_ratio_df is None: continue\n",
    "\n",
    "            # Perform the selected statistical test on the ratios\n",
    "            \n",
    "            # If both tumor and normal only have 0 values for the ratio, the tests don't work\n",
    "            if (sum(tumor_ratio_df['Ratio']) == 0) and (sum(normal_ratio_df['Ratio']) == 0): continue   \n",
    "            # If there are not enough samples, the tests also don't work\n",
    "            if (len(tumor_ratio_df) < 2) or (len(normal_ratio_df) < 2): continue\n",
    "                \n",
    "            if test_type == 'ttest':\n",
    "                test_result = scipy.stats.ttest_ind(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]\n",
    "            else:\n",
    "                # This test throws a warning if the variance in both is 0, so we check for that\n",
    "                if (np.var(tumor_ratio_df['Ratio']) == 0) and (np.var(normal_ratio_df['Ratio']) == 0): continue\n",
    "                test_result = scipy.stats.levene(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]\n",
    "            \n",
    "            # If the p-value is significant, print the result and return the appropriate classification\n",
    "            if test_result < alpha:\n",
    "                sig_result_found = True\n",
    "                # Determine the classification of this ratio\n",
    "                if not (normal_ratio_df['Ratio'] != 0).any():\n",
    "                    num_missing_normal += 1\n",
    "                    print('Normal missing ' + prot_list[i] + ' / ' + prot_list[j])\n",
    "                elif not (tumor_ratio_df['Ratio'] != 0).any():\n",
    "                    num_missing_tumor += 1\n",
    "                    print('Tumor missing ' + prot_list[i] + ' / ' + prot_list[j])\n",
    "    \n",
    "                print(prot_list[i] + ' / ' + prot_list[j] + ': ' + str(test_result))\n",
    "                \n",
    "    if sig_result_found: \n",
    "        \n",
    "        print(complex_name)\n",
    "        print('---------------------------')\n",
    "        print('---------------------------')\n",
    "        \n",
    "        # Classify the complex as a whole depending on the classification of the majority of ratios\n",
    "        # If none of the ratios had all 0 values for tumor or normal, it is classified as \"Other\"\n",
    "        if num_missing_normal == 0 and num_missing_tumor == 0:\n",
    "            return \"Other\"\n",
    "        elif num_missing_normal >= num_missing_tumor:\n",
    "            return \"Normal Missing Proteins\"\n",
    "        elif num_missing_tumor > num_missing_normal:\n",
    "            return \"Tumor Missing Proteins\"\n",
    "        \n",
    "    return \"NS\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### test_all_complexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "test_all_complexes\n",
    "------------------\n",
    "Test all complexes in the dataset for significance\n",
    "\n",
    "Parameters:\n",
    "    test_type (optional) = ttest or levene\n",
    "    \n",
    "Returns:\n",
    "    A dictionary: {\"Normal Missing Proteins\": [List of complexes], \n",
    "    \"Tumor Missing Proteins\": [List of complexes], \n",
    "    \"Other\": [List of complexes]}\n",
    "    \n",
    "\"\"\"\n",
    "\n",
    "def test_all_complexes(test_type = 'ttest'):\n",
    "    \n",
    "    if test_type != 'ttest' and test_type != 'levene':\n",
    "        print(\"Error: test_type must equal 'ttest' or 'levene'\")\n",
    "        return\n",
    "    \n",
    "    classified_complexes = {\"Normal Missing Proteins\": [], \"Tumor Missing Proteins\": [], \"Other\": [], \"NS\": []}\n",
    "    \n",
    "    for key, val in subunitNames.items():\n",
    "        complex_type = test_complex(key, test_type = test_type)\n",
    "        classified_complexes[complex_type].append(key)\n",
    "        \n",
    "    return classified_complexes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### find_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "find_mutations\n",
    "--------------\n",
    "Search mutational data for mutations in proteins that change in a complex or transcription factors\n",
    "known to regulate said proteins\n",
    "\n",
    "Parameters:\n",
    "    complex_dict = dictionary of protein complexes (output of test_all_complexes)\n",
    "    transcription_factors = if True, will test for mutations in transcription factors known to regulate\n",
    "        proteins in the complex\n",
    "        \n",
    "Returns:\n",
    "    Prints results (how many patients have mutations in the protein or transcription factor)\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def find_mutations(complex_dict, transcription_factors = False):\n",
    "    \n",
    "    for key, val in complex_dict.items():\n",
    "        if key == 'NS': continue\n",
    "        print(key)\n",
    "        print('--------------')\n",
    "        print('--------------')\n",
    "        for complex_name in val:\n",
    "            changed_proteins = []\n",
    "            mutation_rate = {}\n",
    "\n",
    "            # Calculate the cutoff for significance (Bonferroni correction)\n",
    "            alpha = 0.05 / (len(subunitNames)*(len(subunitNames) - 1))\n",
    "\n",
    "            # Perform tests to find statistically significant differences in ratios\n",
    "            protein_list = subunitNames[complex_name]\n",
    "            for prot1 in protein_list:\n",
    "                for prot2 in protein_list:\n",
    "                    if prot1 == prot2: continue\n",
    "\n",
    "                    # Test the two proteins (ratio of prot1 / prot2)\n",
    "                    tumor_ratio_df, normal_ratio_df = get_ratio_df(prot1, prot2)\n",
    "                    if tumor_ratio_df is None or normal_ratio_df is None: continue\n",
    "\n",
    "                    # Perform the selected statistical tests on the ratios\n",
    "                    \n",
    "                    # If both tumor and normal only have 0 values for the ratio, the tests don't work\n",
    "                    if (sum(tumor_ratio_df['Ratio']) == 0) and (sum(normal_ratio_df['Ratio']) == 0): continue\n",
    "                    # If there are not enough samples, the tests also don't work\n",
    "                    if (len(tumor_ratio_df) < 2) or (len(normal_ratio_df) < 2): continue\n",
    "                \n",
    "                    ttest_result = scipy.stats.ttest_ind(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]\n",
    "                    # The levene test throws a warning if the variance in both is 0, so we check for that\n",
    "                    if (np.var(tumor_ratio_df['Ratio']) != 0) and (np.var(normal_ratio_df['Ratio']) != 0):\n",
    "                        levene_result = scipy.stats.levene(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]\n",
    "                    else:\n",
    "                        levene_result = 1\n",
    "\n",
    "                    # If the p-value is significant, add the numerator protein to changed_proteins\n",
    "                    if ttest_result < alpha or levene_result < alpha:\n",
    "                        changed_proteins.append(prot1)\n",
    "\n",
    "            changed_proteins = list(set(changed_proteins))\n",
    "            for protein in changed_proteins:\n",
    "                if transcription_factors:\n",
    "                    tfs = tf_data.loc[tf_data['Regulated_Gene'] == protein]['TF']\n",
    "                    for tf in tfs:\n",
    "                        num_mutations = len(set(mutation_data.loc[mutation_data['Gene'] == tf]['Patient_ID']))\n",
    "                        tf_title = tf + ' (regulates ' + protein + ')'\n",
    "                        if num_mutations > 0:\n",
    "                            mutation_rate[tf_title] = num_mutations\n",
    "                else:\n",
    "                    num_mutations = len(set(mutation_data.loc[mutation_data['Gene'] == protein]['Patient_ID']))\n",
    "                    if num_mutations > 0:\n",
    "                        mutation_rate[protein] = num_mutations\n",
    "            if len(mutation_rate) > 0:\n",
    "                print(complex_name)\n",
    "                print(mutation_rate)\n",
    "                print('--------------')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### find_unique_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "find_unique_mutations\n",
    "---------------------\n",
    "Find the number of patients with a mutation in any one of the proteins in a list\n",
    "\n",
    "Parameters:\n",
    "    prot_list = list of proteins to look for mutations in\n",
    "    \n",
    "Returns:\n",
    "    prints the number of patients with a mutation in a protein in prot_list as well as their identifiers\n",
    "    returns a list of those patients\n",
    "\"\"\"\n",
    "\n",
    "def find_unique_mutations(prot_list, cancer_type = 'ov'):\n",
    "    # Set the cancer type\n",
    "    if cancer_type == 'ov': mutation_data = ov_mutation_data\n",
    "    elif cancer_type == 'colon': mutation_data = colon_mutation_data\n",
    "    elif cancer_type == 'en': mutation_data = en_mutation_data\n",
    "    else: print('Error: cancer_type must be \"ov\", \"colon\", or \"en\"')\n",
    "        \n",
    "    patients = []\n",
    "    mutated_proteins = {}\n",
    "    for protein in prot_list:\n",
    "        mutations = list(mutation_data.loc[mutation_data['Gene'] == protein]['Patient_ID'])\n",
    "        if len(mutations) > 0:\n",
    "            mutated_proteins[protein] = len(mutations)\n",
    "            patients = patients + mutations\n",
    "    patients = list(set(patients))\n",
    "    print('Patients with a mutation in any of the ' + str(len(prot_list)) + ' given proteins: ' + str(len(patients)) + '\\n')\n",
    "    print(mutated_proteins)\n",
    "    return patients"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot_ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "plot_ratios\n",
    "-----------\n",
    "Create a seaborn plot for the ratios in a dataframe\n",
    "\n",
    "Parameters:\n",
    "    prot1, prot2 = the two proteins to plot (ratio of prot1 / prot2)\n",
    "    by_patient (optional) = T/F whether or not to visualize matched samples\n",
    "    mutation_list (optional) = plot will distinguish patients with a mutation in any of the proteins in this list\n",
    "    \n",
    "Returns:\n",
    "    Displays a plot of the data\n",
    "    \n",
    "\"\"\"\n",
    "def plot_ratios(prot1, prot2, by_patient = False, mutation_list = None):\n",
    "    \n",
    "    # Set up dataframe for plotting\n",
    "    tumor_ratio_df, normal_ratio_df = get_ratio_df(prot1, prot2)\n",
    "    plot_data = tumor_ratio_df.append(normal_ratio_df)\n",
    "    \n",
    "    # Reformat for visualizing matched samples if necessary\n",
    "    if by_patient or mutation_list:\n",
    "        # Create a new column for matched status\n",
    "        tumor_ratio_df['Matched_Status'] = 'Unmatched'\n",
    "        normal_ratio_df['Matched_Status'] = 'Unmatched'\n",
    "        # Classify samples as matched/unmatched\n",
    "        normal_ratio_df.loc[((normal_ratio_df['Patient_ID']).str.replace('_NM', '')).isin(tumor_ratio_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        tumor_ratio_df.loc[((tumor_ratio_df['Patient_ID']) + '_NM').isin(normal_ratio_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        # Label samples accordingly in plot_data\n",
    "        plot_data = tumor_ratio_df.append(normal_ratio_df)\n",
    "        if mutation_list:\n",
    "            # Find patients with mutations in the given proteins\n",
    "            patient_list = find_unique_mutations(mutation_list)\n",
    "            plot_data.loc[plot_data['Patient_ID'].isin(patient_list), 'Matched_Status'] = 'Has_Mutation'\n",
    "        plot_data.loc[plot_data['Matched_Status'] == 'Unmatched', 'Patient_ID'] = 'Unmatched_Sample'\n",
    "        plot_data['Patient_ID'] = plot_data['Patient_ID'].str.replace('_NM','')\n",
    "    \n",
    "    # Reformat for visualizing a certain list of patients if necessary\n",
    "    if mutation_list and not by_patient:\n",
    "        plot_data.loc[~plot_data['Patient_ID'].isin(patient_list), 'Patient_ID'] = 'Other Mutation'\n",
    "        plot_data.loc[plot_data['Patient_ID'].isin(patient_list), 'Patient_ID'] = 'Given Mutation'\n",
    "    elif mutation_list:\n",
    "        plot_data.loc[plot_data['Patient_ID'].isin(patient_list), 'Patient_ID'] = 'Given Mutation'\n",
    "        \n",
    "    # Print results of statistical tests\n",
    "    print('T-test p-value: ' + str(scipy.stats.ttest_ind(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]))\n",
    "    print('Levene p-value: ' + str(scipy.stats.levene(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]))\n",
    "    \n",
    "    a4_dims = (10, 10)\n",
    "    fig, ax = plt.subplots(figsize=a4_dims)\n",
    "\n",
    "    # Create the plot\n",
    "    if by_patient or patient_list:\n",
    "        boxplt = sns.boxplot(data=plot_data, x='Sample_Type', y='Ratio', color='w', showfliers=False)\n",
    "        boxplt = sns.stripplot(data=plot_data, x='Sample_Type', y='Ratio', hue='Patient_ID', size=10, dodge=True, jitter=True)\n",
    "        boxplt.get_legend().set_bbox_to_anchor((1, 1, 0, 0))\n",
    "    else:\n",
    "        boxplt = sns.boxplot(data=plot_data, x='Sample_Type', y='Ratio', showfliers=False)\n",
    "        boxplt = sns.stripplot(data=plot_data, x='Sample_Type', y='Ratio', dodge=True, jitter=True, color='.3')\n",
    "\n",
    "    # Add styling\n",
    "    boxplt.set_title('Ratio of ' + prot1 + ' / ' + prot2, fontsize='25')\n",
    "    boxplt.set_xlabel('')\n",
    "    boxplt.set_ylabel('Protein Expression Ratio', fontsize='20')\n",
    "    boxplt.tick_params(labelsize='15')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ttest_proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "ttest_proteomics\n",
    "----------------\n",
    "Performs and prints a basic t-test on the proteomics of a given protein (tumor vs normal)\n",
    "\n",
    "Parameters:\n",
    "    protein = protein to test\n",
    "    use_cptac (optional) = T/F whether to use the CPTAC proteomics data (default) or the non-normalized data\n",
    "    \n",
    "Returns:\n",
    "    p-value of the t-test\n",
    "\"\"\"\n",
    "\n",
    "def ttest_proteomics(protein, use_cptac = True, cancer_type = 'ov'):\n",
    "    \n",
    "    # This uses the normalized proteomics data from the CPTAC package\n",
    "    if use_cptac:\n",
    "        # Get data from the appropriate cancer type\n",
    "        if cancer_type == 'ov':\n",
    "            test_data = ov.append_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        elif cancer_type == 'colon': \n",
    "            test_data = colon.append_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        elif cancer_type == 'en': \n",
    "            test_data = en.append_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Proteomics_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        else: print('Error: cancer_type must be \"ov\", \"colon\", or \"en\"')\n",
    "            \n",
    "        test_data = test_data.loc[:, ~test_data.columns.duplicated()]\n",
    "        if cancer_type == 'en':\n",
    "            test_data.rename(columns={protein+'_proteomics': protein, 'Proteomics_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "            test_data.loc[test_data['Sample_Type'] != 'Tumor', 'Sample_Type'] = 'Normal'\n",
    "        else:\n",
    "            test_data.rename(columns={protein+'_proteomics': protein, 'Sample_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "        tumor_df = test_data.loc[test_data['Sample_Type'] == 'Tumor'].dropna(axis = 0)\n",
    "        normal_df = test_data.loc[test_data['Sample_Type'] == 'Normal'].dropna(axis = 0)\n",
    "    \n",
    "    # Otherwise use the non-normalized data to make the plot\n",
    "    else:\n",
    "        test_data = pd.DataFrame(data.loc[data.index == protein].transpose())\n",
    "        test_data['Sample_Type'] = 'Tumor'\n",
    "        test_data.loc[test_data.index.str.contains('_NM'), 'Sample_Type'] = 'Normal'\n",
    "\n",
    "        tumor_df = test_data.loc[test_data['Sample_Type'] == 'Tumor']\n",
    "        normal_df = test_data.loc[test_data['Sample_Type'] == 'Normal']\n",
    "    \n",
    "    # Perform the t-test\n",
    "    return scipy.stats.ttest_ind(tumor_df[protein], normal_df[protein])[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot_phosphoproteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://stackoverflow.com/questions/24685012/pandas-dataframe-renaming-multiple-identically-named-columns\n",
    "# Rename duplicated columns in a dataframe so that each column has a unique name\n",
    "def df_column_uniquify(df):\n",
    "    df_columns = df.columns\n",
    "    new_columns = []\n",
    "    for item in df_columns:\n",
    "        counter = 1\n",
    "        newitem = item\n",
    "        while newitem in new_columns:\n",
    "            counter += 1\n",
    "            newitem = \"{}_{}\".format(item, counter)\n",
    "        new_columns.append(newitem)\n",
    "    df.columns = new_columns\n",
    "    return df\n",
    "\n",
    "def plot_phosphoproteomics(protein, by_patient = False, print_pvals = True, remove_duplicates = False, cancer_type = 'ov'):\n",
    "    \n",
    "    # Get data from the appropriate cancer type\n",
    "    if cancer_type == 'ov':\n",
    "        phos_data = ov.append_metadata_to_omics('clinical', 'phosphoproteomics', \n",
    "                            metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                            omics_genes = protein)\n",
    "    elif cancer_type == 'colon': \n",
    "        phos_data = colon.append_metadata_to_omics('clinical', 'phosphoproteomics', \n",
    "                            metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                            omics_genes = protein)\n",
    "    elif cancer_type == 'en': \n",
    "        phos_data = en.append_metadata_to_omics('clinical', 'phosphoproteomics', \n",
    "                            metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                            omics_genes = protein)\n",
    "    else: print('Error: cancer_type must be \"ov\", \"colon\", or \"en\"')\n",
    "    \n",
    "    if remove_duplicates: phos_data = phos_data.loc[:, ~phos_data.columns.duplicated()]\n",
    "    else: phos_data = df_column_uniquify(phos_data)\n",
    "    plot_data = pd.melt(phos_data, id_vars = ['Patient_ID', 'Sample_Tumor_Normal'], var_name = 'Location', value_name = 'Reading').dropna(axis = 0)\n",
    "    plot_data['Location'] = plot_data['Location'].str.replace('_phosphoproteomics', '')\n",
    "    \n",
    "    # Perform t-tests on tumor vs normal phosphoproteomics for each site\n",
    "    if print_pvals:\n",
    "        for column in phos_data.columns:\n",
    "            if column != 'Patient_ID' and column != 'Sample_Tumor_Normal':\n",
    "                sitedf = phos_data[['Patient_ID', 'Sample_Tumor_Normal', column]]\n",
    "                tumordf = sitedf.loc[sitedf['Sample_Tumor_Normal'] == 'Tumor'].dropna(axis = 0)\n",
    "                normaldf = sitedf.loc[sitedf['Sample_Tumor_Normal'] == 'Normal'].dropna(axis = 0)\n",
    "                if len(tumordf) > 2 and len(normaldf) > 2:\n",
    "                    pval = scipy.stats.ttest_ind(tumordf[column], normaldf[column])[1]\n",
    "                    print(column + ' t-test: ' + str(pval))\n",
    "    \n",
    "    # Plot the data\n",
    "    a4_dims = (20, 20)\n",
    "    fig, ax = plt.subplots(figsize=a4_dims)\n",
    "\n",
    "    boxplt = sns.boxplot(data=plot_data, x='Location', y='Reading', hue='Sample_Tumor_Normal', showfliers=False, palette = 'Blues')\n",
    "    boxplt = sns.stripplot(data=plot_data, x='Location', y='Reading', hue='Sample_Tumor_Normal', dodge=True, jitter=True, color='.3')\n",
    "\n",
    "    # Add styling\n",
    "    boxplt.set_title(protein + ' Phosphorylation', fontsize='25')\n",
    "    boxplt.set_xlabel('')\n",
    "    boxplt.set_ylabel('Phosphorylation Level', fontsize='20')\n",
    "    boxplt.tick_params(labelsize='10')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot_proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "plot_proteomics\n",
    "---------------\n",
    "Create a seaborn plot of the proteomics data for a single protein\n",
    "\n",
    "Parameters:\n",
    "    protein = the protein to plot\n",
    "    use_cptac (optional) = T/F whether to use the CPTAC proteomics data (default) or the non-normalized data\n",
    "    by_patient (optional) = T/F whether or not to visualize matched samples\n",
    "    print_pvals (optional) = T/F whether or not to print the p-values of the statistical tests\n",
    "    \n",
    "Returns:\n",
    "    Displays a plot of the data\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def plot_proteomics(protein, use_cptac = True, by_patient = False, print_pvals = True, cancer_type = 'ov'):\n",
    "    \n",
    "    # This uses the normalized proteomics data from the CPTAC package\n",
    "    if use_cptac:\n",
    "        # Get data from the appropriate cancer type\n",
    "        if cancer_type == 'ov':\n",
    "            plot_data = ov.append_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        elif cancer_type == 'colon': \n",
    "            plot_data = colon.append_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        elif cancer_type == 'en': \n",
    "            plot_data = en.append_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Proteomics_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        else: print('Error: cancer_type must be \"ov\", \"colon\", or \"en\"')\n",
    "            \n",
    "        plot_data = plot_data.loc[:, ~plot_data.columns.duplicated()]\n",
    "        if cancer_type == 'en':\n",
    "            plot_data.rename(columns={protein+'_proteomics': protein, 'Proteomics_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "            plot_data.loc[plot_data['Sample_Type'] != 'Tumor', 'Sample_Type'] = 'Normal'\n",
    "        else:\n",
    "            plot_data.rename(columns={protein+'_proteomics': protein, 'Sample_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "        plot_data = plot_data.dropna(axis = 0)\n",
    "        plot_data['Matched_Status'] = 'Unmatched'\n",
    "        tumor_df = plot_data.loc[plot_data['Sample_Type'] == 'Tumor']\n",
    "        normal_df = plot_data.loc[plot_data['Sample_Type'] == 'Normal']\n",
    "    \n",
    "    # Otherwise use the non-normalized data to make the plot\n",
    "    else:\n",
    "        plot_data = pd.DataFrame(data.loc[data.index == protein].transpose())\n",
    "        plot_data['Sample_Type'] = 'Tumor'\n",
    "        plot_data.loc[plot_data.index.str.contains('_NM'), 'Sample_Type'] = 'Normal'\n",
    "\n",
    "        plot_data['Patient_ID'] = plot_data.index\n",
    "        plot_data['Matched_Status'] = 'Unmatched'\n",
    "        tumor_df = plot_data.loc[plot_data['Sample_Type'] == 'Tumor']\n",
    "        normal_df = plot_data.loc[plot_data['Sample_Type'] == 'Normal']\n",
    "    \n",
    "    # Format to show matched patients if necessary\n",
    "    if by_patient:\n",
    "        pd.options.mode.chained_assignment = None \n",
    "        \n",
    "        # Classify samples as matched/unmatched\n",
    "        normal_df.loc[((normal_df['Patient_ID']).str.replace('_NM|N', '')).isin(tumor_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        tumor_df.loc[((tumor_df['Patient_ID']) + '_NM').isin(normal_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        tumor_df.loc[((tumor_df['Patient_ID']) + 'N').isin(normal_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        tumor_df.loc[('N' + (tumor_df['Patient_ID'])).isin(normal_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        \n",
    "        # Label samples accordingly in plot_data\n",
    "        plot_data = tumor_df.append(normal_df)\n",
    "        plot_data.loc[plot_data['Matched_Status'] == 'Unmatched', 'Patient_ID'] = 'Unmatched_Sample'\n",
    "        plot_data['Patient_ID'] = plot_data['Patient_ID'].str.replace('_NM|N','')\n",
    "    \n",
    "    if print_pvals:\n",
    "        # Print results of statistical tests\n",
    "        print('T-test p-value: ' + str(scipy.stats.ttest_ind(tumor_df[protein], normal_df[protein])[1]))\n",
    "        print('Levene p-value: ' + str(scipy.stats.levene(tumor_df[protein], normal_df[protein])[1]))\n",
    "    \n",
    "    # Create the plot\n",
    "    a4_dims = (10, 10)\n",
    "    fig, ax = plt.subplots(figsize=a4_dims)\n",
    "    \n",
    "    if by_patient:\n",
    "        boxplt = sns.boxplot(data=plot_data, x='Sample_Type', y=protein, color='w', showfliers=False)\n",
    "        boxplt = sns.stripplot(data=plot_data, x='Sample_Type', y=protein, hue='Patient_ID', size=10, dodge=True, jitter=True)\n",
    "        boxplt.get_legend().set_bbox_to_anchor((1, 1, 0, 0))\n",
    "    else:\n",
    "        boxplt = sns.boxplot(data=plot_data, x='Sample_Type', y=protein, showfliers=False)\n",
    "        boxplt = sns.stripplot(data=plot_data, x='Sample_Type', y=protein, dodge=True, jitter=True, color='.3')\n",
    "        \n",
    "    # Add styling\n",
    "    boxplt.set_title(protein + ' Proteomics', fontsize='25')\n",
    "    boxplt.set_xlabel('')\n",
    "    boxplt.set_ylabel('Protein Expression', fontsize='20')\n",
    "    boxplt.tick_params(labelsize='15')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### tidy_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Thanks StackOverflow! \n",
    "# https://stackoverflow.com/questions/12680754/split-explode-pandas-dataframe-string-entry-to-separate-rows\n",
    "\n",
    "def tidy_split(df, column, sep='|', keep=False):\n",
    "    \"\"\"\n",
    "    Split the values of a column and expand so the new DataFrame has one split\n",
    "    value per row. Filters rows where the column is missing.\n",
    "\n",
    "    Params\n",
    "    ------\n",
    "    df : pandas.DataFrame\n",
    "        dataframe with the column to split and expand\n",
    "    column : str\n",
    "        the column to split and expand\n",
    "    sep : str\n",
    "        the string used to split the column's values\n",
    "    keep : bool\n",
    "        whether to retain the presplit value as it's own row\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pandas.DataFrame\n",
    "        Returns a dataframe with the same columns as `df`.\n",
    "    \"\"\"\n",
    "    indexes = list()\n",
    "    new_values = list()\n",
    "    #df = df.dropna(subset=[column])\n",
    "    for i, presplit in enumerate(df[column].astype(str)):\n",
    "        values = presplit.split(sep)\n",
    "        if keep and len(values) > 1:\n",
    "            indexes.append(i)\n",
    "            new_values.append(presplit)\n",
    "        for value in values:\n",
    "            indexes.append(i)\n",
    "            new_values.append(value)\n",
    "    new_df = df.iloc[indexes, :].copy()\n",
    "    new_df[column] = new_values\n",
    "    return new_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
